### PR TITLE
Fix huge lag spike on new combo in osu!catch

### PR DESCRIPTION
**Pull request by Terochi**
Addresses https://github.com/ppy/osu/issues/23220

Makes use of already allocated `PoolableDrawable`s instead of allocating new ones.
I could probably imagine some issues with this approach, but not sure.

Also fixes dropped objects not being removed when rewinding (at the end of the video).

https://user-images.githubusercontent.com/100224757/233731777-5d373522-cc89-416b-85ec-897b95fcfe66.mp4

https://user-images.githubusercontent.com/100224757/233731757-5816f69d-a0ee-4cf8-a04d-bb82b1cdec7f.mp4

### CODE REVIEW

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 1]
```diff
@@ -430,13 +433,11 @@ private void clearPlate(DroppedObjectAnimation animation)
...
         {
-            var droppedObject = getDroppedObject(caughtObject);
-
             caughtObjectContainer.Remove(caughtObject, false);
 
-            droppedObjectTarget.Add(droppedObject);
+            droppedObjectTarget.Add(getDroppedObject(caughtObject));
 
-            applyDropAnimation(droppedObject, animation);
+            applyDropAnimation(caughtObject, animation);
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:440]: Readability wise this doesn't make muich sense to me. Why is the drop animation applied to the original object here?

Comment (Terochi) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:440]: Oh I completely missed this, my bad.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 2]
```diff
@@ -418,10 +418,13 @@ private CaughtObject getDroppedObject(CaughtObject caughtObject)
 
         private void clearPlate(DroppedObjectAnimation animation)
         {
-            var droppedObjects = caughtObjectContainer.Children.Select(getDroppedObject).ToArray();
+            var caughtObjects = caughtObjectContainer.Children.ToArray();
 
             caughtObjectContainer.Clear(false);
 
+            // Use the already returned PoolableDrawables for new objects
+            var droppedObjects = caughtObjects.Select(getDroppedObject).ToArray();
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: This feels sorta dangerous. I suppose the intent here is to first return the caught drawable objects to pool, so that they can be reused for the drop. But there is no strong guarantee that something else hasn't applied a new hitobject to the caught DHOs in the meantime after the DHOs have been returned to pool. So you could potentially end up dropping an object that was never dropped.

This seems salvageable if you dump `caughtObjectContainer.Children.Select(dho => dho.HitObject).ToArray()` instead rather than the DHOs raw. That eliminates any risk of a mix-up.

Comment (Terochi) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: I was thinking there would be some issue of this sort with this approach, still I'm not sure that situation could happen. Will adjust anyway just in case.

Comment (Terochi) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: Not actually sure how to go about this, as I would need to save stuff from `IHasCatchObjectState` and also the objects last position, which would look messy. Correct me if I'm wrong, but the scenario that you are proposing would only potentially happen if multiple objects were caught at the same time, which won't happen?

Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: > the scenario that you are proposing would only potentially happen if multiple objects were caught at the same time

I'm not sure where this assertion is coming from. It doesn't sound correct. Objects scrolling onto the screen will also retrieve objects from the pool.

Is this change even doing much work? I would wager that most of the gains here actually come from the lifetime set.

Comment (Terochi) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: > Objects scrolling onto the screen will also retrieve drawable objects from the pool.

Those are not part of the `Catcher` class. Here, only the objects placed on the catcher are being allocated.

> Is this change even doing much work? I would wager that most of the gains here actually come from the lifetime set.

When comparing the videos it seems to cut the allocated memory and draw calls in half.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 3]
```diff
@@ -430,10 +433,10 @@ private void clearPlate(DroppedObjectAnimation animation)
 
         private void removeFromPlate(CaughtObject caughtObject, DroppedObjectAnimation animation)
         {
-            var droppedObject = getDroppedObject(caughtObject);
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:433]: Why was this line moved? Is there meaning behind it? If it's required, it's very confusing as to why, to the point I question the rest of the change.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:426]: I think this is safe enough, based on the above assumption (the pools are local and usages are all in the method used here).

File: osu.Game.Rulesets.Catch/UI/Catcher.cs (hunk 3 above)
Comment (Terochi) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:433]: It's the same logic as with `clearPlate`. It definitely isn't required, as you won't really notice the optimization for a single object.

Comment (peppy) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:433]: Right, fair enough.

**Review by peppy — APPROVED** [review]
Gonna get this in. I think it's okay.
